### PR TITLE
Build Buster rootfs using kci_rootfs

### PR DESCRIPTION
**Pull request by Lakshmipathi**
Use kci_rootfs in Jenkins job to create required rootfs for given architecture.

Signed-off-by: Lakshmipathi Ganapathi <lakshmipathi.ganapathi@collabora.co.uk>

### CODE REVIEW

File: jenkins/rootfs-builder.jpl [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  This library is distributed in the hope that it will be useful, but WITHOUT
+  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+  FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+  details.
+
+  You should have received a copy of the GNU Lesser General Public License
+  along with this library; if not, write to the Free Software Foundation, Inc.,
+  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+*/
+
```
Comment (mgalka) [inline jenkins/rootfs-builder.jpl]: Please, document job parameters

**Comment by Lakshmipathi** [comment]
Passed on staging server.

**Comment by gctucker** [comment]
> Passed on staging server.

Please provide some links to the Jenkins job logs and artifacts.

Also please create a PR in `kernelci-jenkins` to create the new rootfs jobs, they will be added automatically in the next staging run (evey 8h).

**Comment by Lakshmipathi** [comment]
> Passed on staging server.
URL: https://bot.staging.kernelci.org/job/laks/job/rootfs-builder/

**Comment by gctucker** [comment]
> > Passed on staging server.
> > URL: https://bot.staging.kernelci.org/job/laks/job/rootfs-builder/

This was a buster-igt job:
```
./kci_rootfs build --config buster-igt --arch amd64 --data-path jenkins/debian/debos
```

We need to have the trigger job and the buster rootfs generated with the jobs now, with the new jobs created by kernelci-jenkins.

**Comment by Lakshmipathi** [comment]
Above job has results for buster, buster-igt rootfs builds.  Buster: https://bot.staging.kernelci.org/job/laks/job/rootfs-builder/1/console
```
+ ./kci_rootfs build --config buster --arch amd64 --data-path jenkins/debian/debos
```
Buster-igt: https://bot.staging.kernelci.org/job/laks/job/rootfs-builder/4/console

```
+ ./kci_rootfs build --config buster-igt --arch amd64 --data-path jenkins/debian/debos
```

**Comment by Lakshmipathi** [comment]
Now made architecture (ROOTFS_ARCH) as a mandatory parameter.

File: jenkins/rootfs-trigger.jpl [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+""", returnStdout: true).trim()
+
+    def rootfs_config_list =  rootfs_config_list_raw.tokenize('\n')
+
+    for (String rootfs_config_raw: rootfs_config_list) {
+        def data = rootfs_config_raw.tokenize(' ')
+        def config = data[0]
+        def arch = data[1]
+        config_list.add([config, arch])
+    }
```
Comment (gctucker) [inline jenkins/rootfs-trigger.jpl]: Indentation problem.

File: jenkins/rootfs-builder.jpl [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  Configuration name to build RootFS images
+ROOTFS_ARCH
+  RootFS image architecture
+*/
+
+
+@Library('kernelci') _
+import org.kernelci.util.Job
+
+def build_rootfs(config, arch, pipeline_version,kci_core) {
```
Comment (gctucker) [inline jenkins/rootfs-builder.jpl]: Missing space before `kci_core`.

File: jenkins/rootfs-builder.jpl [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    def docker_image = "${params.DOCKER_BASE}debos"
+    def pipeline_version =  "${params.PIPELINE_VERSION}"
+
+    j.dockerPullWithRetry(docker_image).inside() {
+        j.cloneKciCore(kci_core, params.KCI_CORE_URL, params.KCI_CORE_BRANCH)
+    }
+
+    j.dockerPullWithRetry(docker_image).inside(" --privileged --device /dev/kvm ") {
+        stage("Build ${config} ${arch}") {
+            timeout(time: 60, unit: 'MINUTES') {
```
Comment (gctucker) [inline jenkins/rootfs-builder.jpl]: Please increase the timeout to 4h i.e. 240 minutes.  Some test suites may take a very long time to build.

File: jenkins/rootfs-trigger.jpl [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+    j.dockerPullWithRetry(docker_image).inside() {
+
+        stage("Init") {
+            timeout(time: 15, unit: 'MINUTES') {
+                j.cloneKciCore(
+                    kci_core, params.KCI_CORE_URL, params.KCI_CORE_BRANCH)
+            }
+
```
Comment (gctucker) [inline jenkins/rootfs-trigger.jpl]: Superfluous blank line.

File: jenkins/rootfs-trigger.jpl [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        def arch = data[1]
+        config_list.add([config, arch])
+    }
+    }
+}
+
+
+def buildRootfsStep(job, config ,arch) {
+    def node_label = "docker"
+    def parallel_builds = "4"
```
Comment (gctucker) [inline jenkins/rootfs-trigger.jpl]: These variabes aren't used anywhere, they seem to have been copy-pasted from `build-trigger.jpl`.

**Comment by Lakshmipathi** [comment]
Added above listed changes.

File: jenkins/rootfs-builder.jpl [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    withCredentials([string(credentialsId: params.KCI_TOKEN_ID, variable: 'API_TOKEN')]) {
+            sh(script: """\
+./kci_rootfs \
+upload \
+--token ${API_TOKEN} \
+--api ${params.KCI_API_URL} \
+--rootfs-dir ${pipeline_version} \
+--upload-path images/rootfs/debian/${config}/${pipeline_version}
+""")
+
```
Comment (gctucker) [inline jenkins/rootfs-builder.jpl]: Superfluous blank line.

File: jenkins/rootfs-builder.jpl [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        }
+    }
+
```
Comment (gctucker) [inline jenkins/rootfs-builder.jpl]: Superfluous blank line.

File: jenkins/rootfs-trigger.jpl [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+
+                builds[step_name] = buildRootfsStep(
+                    "rootfs-builder", config_name, arch)
+
+                i += 1
+            }
+
+            parallel(builds)
+        }
+
```
Comment (gctucker) [inline jenkins/rootfs-trigger.jpl]: Superfluous blank line.

**Comment by gctucker** [comment]
@Lakshmipathi The last staging build for this was last week, can you please run a new one to check the result is still fine with the changes made since then?

**Comment by Lakshmipathi** [comment]
Updated and tested on staging https://bot.staging.kernelci.org/job/laks/job/rootfs-builder/7/

**Comment by gctucker** [comment]
This still needs to be tested.  It's been very close to getting merged for several weeks now, just lacking the final work to test and wrap things up.  We need to resolve it for the next release.

**Comment by Lakshmipathi** [comment]
Tested buster rootfs with LAVA and here are the results:

buster/armel job: https://lava.collabora.co.uk/scheduler/job/2320520/definition
buster/arm64 job: https://lava.collabora.co.uk/scheduler/job/2320918/definition
buster/i386 job: https://lava.collabora.co.uk/scheduler/job/2320941/definition
buster/amd64 job: https://lava.collabora.co.uk/scheduler/job/2321009/definition

**Comment by gctucker** [comment]
Testing on staging with fixes from #355.

**Pull request merged**